### PR TITLE
Remove redirect_uri Validations for request_uri Authorization Flows

### PR DESCRIPTION
**Pull request by thiloshon**
Resolves https://github.com/wso2/product-is/issues/9401

Description

PAR spec request expects `redirect_uri` to be omitted from the authorization call [1] as it is already added to the request object through PAR endpoint. But currently any requests without `redirect_uri` fails in IAM. This PR removes the check for `redirect_uri` when authorization call comes with `request_uri`

[1] https://tools.ietf.org/html/draft-ietf-oauth-par-03#section-4

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java [hunk 1]
```diff
@@ -58,6 +64,10 @@ private static boolean isContainOIDCScope(String scope) {
     @Override
     public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
 
+        if (!StringUtils.isBlank(request.getParameter(REQUEST_URI))) {
+            requiredParams = new ArrayList<>(Arrays.asList(CLIENT_ID, RESPONSE_TYPE, SCOPE, REQUEST_URI));
+            notAllowedParams.add(REQUEST);
+        }
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:70]: Can you explain the reasoning behind this. It looks like we are marking the required params based on external input. IMO we shouldn't do validations based on external input. these make spaces for loopholes.

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java [hunk 2]
```diff
@@ -58,6 +64,10 @@ private static boolean isContainOIDCScope(String scope) {
     @Override
     public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
 
+        if (!StringUtils.isBlank(request.getParameter(REQUEST_URI))) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:67]:
```suggestion
        if (StringUtils.isNotBlank(request.getParameter(REQUEST_URI))) {
```

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java [hunk 3]
```diff
@@ -58,6 +64,10 @@ private static boolean isContainOIDCScope(String scope) {
...
+            requiredParams = new ArrayList<>(Arrays.asList(CLIENT_ID, RESPONSE_TYPE, SCOPE, REQUEST_URI));
+            notAllowedParams.add(REQUEST);
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:69]: What where the initial values for these?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 4]
```diff
@@ -681,6 +683,11 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
         String sessionDataKeyConsent = UUID.randomUUID().toString();
         try {
             if (entry != null && entry.getQueryString() != null) {
+
+                if (entry.getQueryString().contains(REQUEST_URI) && params != null) {
+                    entry.setQueryString(entry.getQueryString() +
+                            "&" + PROP_REDIRECT_URI + "=" + params.getRedirectURI());
+                }
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:690]: Why is this needed? Also if `PROP_REDIRECT_URI` is required downstream, then what would happen if the App is configured to skip consent?(IIRC then this code block will not execute)

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java (hunk 3 above)
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:69]: Its `response_type`, `client_id` and `redirect_uri`. For PAR we need these four as mandatory and `request` shouldnt be there.

File: components/org.wso2.carbon.identity.oauth.common/src/test/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidatorTest.java [hunk 5]
```diff
@@ -95,6 +95,69 @@ public void testValidateRequiredParameters(Map<String, String> headerMap, boolea
         }
     }
 
+    @DataProvider(name = "Request Provider Request Uri")
+    public Object[][] getRequestParamsForRequestUri() {
+
+        Map<String, String> validUriRequest = new HashMap<>();
+        validUriRequest.put(SCOPE, OAuthConstants.Scope.OPENID);
+        validUriRequest.put("response_type", getResponseTypeValue());
+        validUriRequest.put("request_uri", "company:domain:urn");
```
[the remark this answers is not in the document]
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.common/src/test/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidatorTest.java:104]: Done

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java [hunk 6]
```diff
@@ -137,7 +138,8 @@ private void validateInputParameters(OAuthMessage oAuthMessage) throws InvalidRe
                     OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ErrorCodes.OAuth2SubErrorCodes.INVALID_CLIENT);
         }
 
-        if (StringUtils.isBlank(oAuthMessage.getRequest().getParameter(REDIRECT_URI))) {
+        if (StringUtils.isBlank(oAuthMessage.getRequest().getParameter(REQUEST_URI)) &&
```
[the remark this answers is not in the document]
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java:141]: The request object is already validate as part of/par endpoint. If the request object doest contain redirect_uri par call will end in error and the flow cannot go further.

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:69]: scope is not a mandatory param in oauth/oidc why is it added here?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java:141]: request_uri will endup creating a request object in the oauth component as well. so does the request object in the oauth component have a validation for this? Also according to the spec is the redirect_uri a mandatory param for the request object?

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java (hunk 3 above)
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:69]: I added as it is required in OB spec, but I understand it is not OIDC compliant. I will remove it.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java:141]: the creation of request object from request_uri is delegated to Object Builder extension. Currently we are maintaining a builder for OB use cases in OB. If you require a validation or even not have one, you can add that in the IS specific extension implementation. Since you have that control, you can validate redirect_uri as per IS requirements too.

[the remark this answers is not in the document]
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/state/OAuthRequestStateValidator.java:141]: Fixed!

**Review by janakamarasena — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java [hunk 7]
```diff
@@ -58,6 +64,10 @@ private static boolean isContainOIDCScope(String scope) {
     @Override
     public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
 
+        if (StringUtils.isNotBlank(request.getParameter(REQUEST_URI))) {
+            requiredParams = new ArrayList<>(Arrays.asList(CLIENT_ID, RESPONSE_TYPE, REQUEST_URI));
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:68]: A small comment as to why we do this for later reference would be nice

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 8]
```diff
@@ -1686,6 +1690,13 @@ private void handleRequestObject(OAuthMessage oAuthMessage, OAuthAuthzRequest oa
              */
         overrideAuthzParameters(oAuthMessage, parameters, oauthRequest.getParam(REQUEST),
                 oauthRequest.getParam(REQUEST_URI), requestObject);
+
+        // If the redirect uri was not given in auth request the registered redirect uri will be available here,
+        // so validating if the registered redirect uri is a single uri that can be properly redirected.
+        if (parameters.getRedirectURI().startsWith(REGEX_PATTERN)) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1696]: Can we guarantee parameters.getRedirectURI() won't be nulll?

Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1696]: RedirectURI is populated in `populateOauthParameters()` method. If it is null, it will be captured here itself.

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java (hunk 7 above)
Comment (thiloshon) [reply components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:70]: Fixed!

File: components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java [hunk 9]
```diff
@@ -58,6 +64,12 @@ private static boolean isContainOIDCScope(String scope) {
     @Override
     public void validateRequiredParameters(HttpServletRequest request) throws OAuthProblemException {
 
+        if (StringUtils.isNotBlank(request.getParameter(REQUEST_URI))) {
+            // PAR spec mandates request_uri to have client_id, response_type in accordance to OAuth 2.0
+            // Also request is disallowed if request_uri is used.
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.common/src/main/java/org/wso2/carbon/identity/oauth/common/CodeTokenResponseValidator.java:69]:
```suggestion
            // Also 'request' parameter is disallowed if 'request_uri' parameter is present in the authorization request.
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 10]
```diff
@@ -681,6 +683,11 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
         String sessionDataKeyConsent = UUID.randomUUID().toString();
         try {
             if (entry != null && entry.getQueryString() != null) {
+
+                if (entry.getQueryString().contains(REQUEST_URI) && params != null) {
+                    entry.setQueryString(entry.getQueryString() +
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:688]: Add a comment here as to why we do this please :)

**Review by mefarazath — APPROVED** [review]

**Review by mefarazath — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 11]
```diff
@@ -1686,6 +1690,13 @@ private void handleRequestObject(OAuthMessage oAuthMessage, OAuthAuthzRequest oa
              */
         overrideAuthzParameters(oAuthMessage, parameters, oauthRequest.getParam(REQUEST),
                 oauthRequest.getParam(REQUEST_URI), requestObject);
+
+        // If the redirect uri was not given in auth request the registered redirect uri will be available here,
+        // so validating if the registered redirect uri is a single uri that can be properly redirected.
+        if (StringUtils.startsWith(parameters.getRedirectURI(), REGEX_PATTERN)) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1696]:
```suggestion
        if (StringUtils.isBlank(parameters.getRedirectURI()) || StringUtils.startsWith(parameters.getRedirectURI(), REGEX_PATTERN)) {
```
lets add this as, if the redirect_uri is null/blank then there should be a error

**Review by janakamarasena — APPROVED** [review]
